### PR TITLE
 KafkaSink Visualization

### PR DESCRIPTION
**Pull request by sahil143**
**Fixes**: https://issues.redhat.com/browse/ODC-6628

**Analysis / Root cause**: Missing visualization of kafka Sink

**Solution Description**: Added support for Kafka sink in Topology view

**Screen shots / Gifs for design review**: 

Kapture 2022-04-11 at 20 29 55

**Unit test coverage report**: 

**Test setup:**

Creation of Kafka Sink

- Install Serverless Operator
- Create KnativeKafka CR (KnativeEventing CR is required to create KF CR). 
- Expand Sink section on Form view and enable sink
- Install AMQ streams
- Create namespace kafka
- Create Kaka cluster (Kafka CR from AMQ streams) with default values (make note of name/ns for example- my-cluster-kafka-bootstrap.kafka.svc:9092)
- Create kafka topic as below
```
apiVersion: kafka.strimzi.io/v1beta2
kind: KafkaTopic
metadata:
  name: knative-demo-topic
  namespace: kafka
  labels:
    strimzi.io/cluster: my-cluster # kafka cluster name
spec:
  partitions: 3
  replicas: 1
  config:
    retention.ms: 7200000
    segment.bytes: 1073741824
```
- Create Kafka Sink
```
apiVersion: eventing.knative.dev/v1alpha1
kind: KafkaSink
metadata:
  name: my-kafka-sink
  namespace: kafka
spec:
  topic: knative-demo-topic # KafkaTopic name
  bootstrapServers:
   - my-cluster-kafka-bootstrap.kafka:9092 # Kafka cluster bootstrap server
 ```

**Browser conformance**: 

- [ ] Chrome
- [ ] Firefox
- [ ] Safari
- [ ] Edge

### CODE REVIEW

File: frontend/packages/knative-plugin/src/topology/knative-topology-utils.ts [hunk 1]
```diff
@@ -1159,6 +1159,14 @@ export const transformKnNodeData = (
...
     const item = createKnativeDeploymentItems(res, resources, utils);
     switch (type) {
+      case NodeType.KafkaSink: {
+        const data = createEventSinkTopologyNodeData(res, item, type);
+        const itemData = getOwnedEventSinkData(res, data, resources);
+        knDataModel.nodes.push(...getKnativeTopologyNodeItems(res, type, itemData, resources));
+        const newGroup = getTopologyGroupItems(res);
+        mergeGroup(newGroup, knDataModel.nodes);
+        break;
+      }
```
Comment (invincibleJai) [inline frontend/packages/knative-plugin/src/topology/knative-topology-utils.ts:1169]: can we have them in same case? adn i blv for `kafkaSink` `getEventSinkTopologyEdgeItems` would return `[]`. WDYT?
```suggestion
      case NodeType.EventSink:
      case NodeType.KafkaSink: {
        const data = createEventSinkTopologyNodeData(res, item, type);
        const itemData = getOwnedEventSinkData(res, data, resources);
        knDataModel.nodes.push(...getKnativeTopologyNodeItems(res, type, itemData, resources));
        const newGroup = getTopologyGroupItems(res);
        mergeGroup(newGroup, knDataModel.nodes);
        break;
      }
```

**Review by karthikjeeyar — COMMENTED** [review]
Edit application group action is missing in kakfa sink node. I am able to drag and drop to a different application group, so I think we need to add edit application group option in context menu.

image



Deleting an application group doesn't remove Kafka sink resource

image

**Review by karthikjeeyar — COMMENTED** [review]
@sahil143 Since we do not have any underlying resources In sidebar we show only detail tab, and this is expected right?

File: frontend/packages/knative-plugin/src/topology/data-transformer.ts [hunk 2]
```diff
@@ -38,17 +38,37 @@ const addKnativeTopologyData = (
   addToTopologyDataModel(knativeResourceDataModel, graphModel);
 };
 
+const getKnativeTOpologyDataUtils = () => [
```
Comment (karthikjeeyar) [inline frontend/packages/knative-plugin/src/topology/data-transformer.ts:41]: nit:
```suggestion
const getKnativeTopologyDataUtils = () => [
```

File: frontend/packages/topology/src/utils/application-utils.ts [hunk 3]
```diff
@@ -237,6 +239,10 @@ export const cleanUpWorkload = async (
     case KnativeServiceModel.kind:
       batchDeleteRequests(knativeDeleteModels, resourceData);
       break;
+    case CamelKameletBindingModel.kind:
+    case KafkaSinkModel.kind:
+      deleteRequest(resourceModel, resource);
+      break;
```
Comment (karthikjeeyar) [inline frontend/packages/topology/src/utils/application-utils.ts:245]: Can you add test for this application-utils.

**Comment by karthikjeeyar** [comment]
@sahil143 Code changes looks good to me, started a clusterbot to test again. 
It would be great if we can add unit test case wherever applicable.

File: frontend/packages/topology/src/utils/__tests__/application-utils.spec.ts [hunk 4]
```diff
@@ -172,6 +175,27 @@ describe('ApplicationUtils ', () => {
       .catch((err) => fail(err));
   });
 
+  it('Should delete all the specific models related to kamelet binding', async (done) => {
```
Comment (karthikjeeyar) [inline frontend/packages/topology/src/utils/__tests__/application-utils.spec.ts:178]: can we add similar test for the new KafkaSink as well?  I see you have added test data `sampleSourceKafkaSink` but no see any tests related to it 🤔

Comment (sahil143) [reply frontend/packages/topology/src/utils/__tests__/application-utils.spec.ts:201]: fixed

File: frontend/packages/topology/src/utils/application-utils.ts (hunk 3 above)
Comment (sahil143) [reply frontend/packages/topology/src/utils/application-utils.ts:245]: done

**Review by vikram-raj — COMMENTED** [review]
Verified the changes on the cluster bot. And it works as expected.

/lgtm
/label qe-approved

**Pull request merged**